pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_alter_primary_key.go

```diff
@@ -0,0 +1,797 @@
+				"Column element.", col.Column))
+		}
+		if colCurrentStatus == scpb.Status_DROPPED || colCurrentStatus == scpb.Status_ABSENT {
+			panic(pgerror.Newf(pgcode.ObjectNotInPrerequisiteState,
+				"column %q is being dropped", col.Column))
+		}
+		if colElem.IsInaccessible {
+			panic(pgerror.Newf(pgcode.InvalidSchemaDefinition, "cannot use inaccessible "+
+				"column %q in primary key", col.Column))
+		}
+		_, _, colTypeElem := scpb.FindColumnType(colElems)
+		if colTypeElem == nil {
+			panic(fmt.Sprintf("programming error: resolving column %v does not give a "+
+				"ColumnType element.", col.Column))
+		}
+		if colTypeElem.IsNullable {
+			panic(pgerror.Newf(pgcode.InvalidSchemaDefinition, "cannot use nullable column "+
+				"%q in primary key", col.Column))
+		}
+
+		if !b.EvalCtx().Settings.Version.IsActive(b, clusterversion.Start22_1) {
+			if colTypeElem.IsVirtual {
+				panic(pgerror.Newf(pgcode.FeatureNotSupported, "cannot use virtual column %q "+
+					"in primary key", col.Column))
+			}
+		}
+	}
+}
+
+// isNewPrimaryKeySameAsOldPrimaryKey returns whether the requested new
+// primary key is the same as the old primary key.
+func isNewPrimaryKeySameAsOldPrimaryKey(
+	b BuildCtx, tbl *scpb.Table, t *tree.AlterTableAlterPrimaryKey,
+) bool {
+	oldPrimaryIndexElem := mustRetrievePrimaryIndexElement(b, tbl.TableID)
+	oldPrimaryIndexKeyColumns := mustRetrieveKeyIndexColumns(b, tbl.TableID, oldPrimaryIndexElem.IndexID)
+
+	// check whether they have the same number of key columns.
+	if len(oldPrimaryIndexKeyColumns) != len(t.Columns) {
+		return false
+	}
+
+	// check whether they are both sharded or both not sharded.
+	if (oldPrimaryIndexElem.Sharding == nil) != (t.Sharded == nil) {
+		return false
+	}
+
+	// check whether all key columns (ID and directions) are the same.
+	for i, col := range t.Columns {
+		colElems := b.ResolveColumn(tbl.TableID, col.Column, ResolveParams{
+			IsExistenceOptional: false,
+			RequiredPrivilege:   privilege.CREATE,
+		})
+		_, _, colElem := scpb.FindColumn(colElems)
+		if (oldPrimaryIndexKeyColumns[i].ColumnID != colElem.ColumnID) ||
+			oldPrimaryIndexKeyColumns[i].Direction != col.Direction.ToIndexColumnDirection() {
+			return false
+		}
+	}
+
+	// If both are sharded, check whether they have the same bucket count.
+	if oldPrimaryIndexElem.Sharding != nil {
+		shardBucketsInNewPrimaryIndex, err := tabledesc.EvalShardBucketCount(b, b.SemaCtx(), b.EvalCtx(),
+			t.Sharded.ShardBuckets, t.StorageParams)
+		if err != nil {
+			panic(err)
+		}
+		if oldPrimaryIndexElem.Sharding.ShardBuckets != shardBucketsInNewPrimaryIndex {
+			return false
+		}
+	}
+
+	return true
+}
+
+// fallBackIfConcurrentSchemaChange panics with an unimplemented error if
+// there are any other concurrent schema change on this table. This is determined
+// by searching for any element that is in ABSENT (resp. PUBLIC) and targeting
+// PUBLIC (resp. ABSENT).
+func fallBackIfConcurrentSchemaChange(b BuildCtx, tableID catid.DescID) {
```

Comment:
I don't think this is good enough. Consider the case where we've had a previous statement in the same transaction but all of the elements in the statement phase have moved away from their initial status. Consider just checking if the elements are not in their terminal state.